File: NRDB/CardList.m
//
//  CardList.m
//  NRDB
//
//  Created by Gereon Steffens on 29.11.13.
//  Copyright (c) 2014 Gereon Steffens. All rights reserved.
//

#import "CardList.h"
#import "Card.h"
#import "Deck.h"
#import "CardSets.h"
#import "CardManager.h"

@interface CardList()

@property NRRole role;
@property NSMutableArray* initialCards;

@property int cost;
@property NSString* type;
@property NSSet* types;
@property NSString* subtype;
@property NSSet* subtypes;
@property int strength;
@property int mu;
@property int trash;
@property NSString* faction;
@property NSSet* factions;
@property int influence;
@property NSString* set;
@property NSSet* sets;
@property int agendaPoints;
@property NSString* text;
@property NRSearchScope searchScope;
@property BOOL unique;
@property BOOL limited;
@property BOOL altart;
@property NRFaction faction4inf;    // faction for influence filter

@property NRCardListSortType sortType;

@end

@implementation CardList

-(CardList*) initForRole:(NRRole)role
{
    if ((self = [super init]))
    {
        self.role = role;
        self.sortType = NRCardListSortA_Z;
        self.faction4inf = NRFactionNone;
        [self resetInitialCards];
        [self clearFilters];
    }
    return self;
}

+(CardList*) browserInitForRole:(NRRole)role
{
    CardList* cl = [[CardList alloc] init];
    cl.sortType = NRCardListSortA_Z;
    cl.role = role;
    
    NSArray* roles;
    switch (role)
    {
        case NRRoleNone:
            roles = @[ @(NRRoleRunner), @(NRRoleCorp) ];
            break;
        case NRRoleCorp:
            roles = @[ @(NRRoleCorp) ];
            break;
        case NRRoleRunner:
            roles = @[ @(NRRoleRunner) ];
            break;
    }
    
    cl.initialCards = [NSMutableArray array];
    
    for (NSNumber* r in roles)
    {
        [cl.initialCards addObjectsFromArray:[CardManager allForRole:r.intValue]];
        [cl.initialCards addObjectsFromArray:[CardManager identitiesForRole:r.intValue]];
    }
    [cl filterDeselectedSets];
    [cl clearFilters];
    
    return cl;
}

-(void) resetInitialCards
{
    self.initialCards = [NSMutableArray arrayWithArray:[CardManager allForRole:self.role]];
    [self filterDeselectedSets];
}

-(void) filterDeselectedSets
{
    // remove all cards from sets that are deselected
    NSSet* disabledSetCodes = [CardSets disabledSetCodes];
    NSPredicate* predicate = [NSPredicate predicateWithFormat:@"!(setCode in %@)", disabledSetCodes];
    [self.initialCards filterUsingPredicate:predicate];
}

-(void) filterAgendas:(Card *)identity
{
    [self resetInitialCards];
    
    if (identity && identity.faction != NRFactionNeutral)
    {
        NSArray* factions = @[ @(NRFactionNeutral), @(identity.faction) ];
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"type != %d OR (type = %d AND faction in %@)", NRCardTypeAgenda, NRCardTypeAgenda, factions];
        
        [self.initialCards filterUsingPredicate:predicate];
    }
    
    [self applyFilters];
}

-(void) clearFilters
{
    self.cost = -1;
    self.type = @"";
    self.types = nil;
    self.subtype = @"";
    self.subtypes = nil;
    self.strength = -1;
    self.mu = -1;
    self.trash = -1;
    self.faction = @"";
    self.factions = nil;
    self.influence = -1;
    self.set = @"";
    self.sets = nil;
    self.agendaPoints = -1;
    self.text = @"";
    self.searchScope = NRSearchAll;
    self.unique = NO;
    self.limited = NO;
    self.altart = NO;
    self.faction4inf = NRFactionNone;
}

-(void) filterByType:(NSString*) type
{
    self.type = type;
    self.types = nil;
}

-(void) filterByTypes:(NSSet *)types
{
    self.type = @"";
    self.types = types;
}

-(void) filterByFaction:(NSString*) faction
{
    self.faction = faction;
    self.factions = nil;
}

-(void) filterByFactions:(NSSet*) factions
{
    self.faction = @"";
    self.factions = factions;
}

-(void) filterByText:(NSString*) text;
{
    self.text = text;
    self.searchScope = NRSearchText;
}

-(void) filterByTextOrName:(NSString*) text;
{
    self.text = text;
    self.searchScope = NRSearchAll;
}

-(void) filterByName:(NSString*) name;
{
    self.text = name;
    self.searchScope = NRSearchName;
}

-(void) filterBySet:(NSString*) set;
{
    self.set = set;
    self.sets = nil;
}

-(void) filterBySets:(NSSet*) sets;
{
    self.set = @"";
    self.sets = sets;
}

-(void) filterByInfluence:(int)influence
{
    self.influence = influence;
    self.faction4inf = NRFactionNone;
}

-(void) filterByInfluence:(int)influence forFaction:(NRFaction)faction
{
    self.influence = influence;
    self.faction4inf = faction;
}

-(void) filterByMU:(int)mu
{
    self.mu = mu;
}

-(void) filterByTrash:(int)trash
{
    self.trash = trash;
}

-(void) filterByCost:(int)cost
{
    self.cost = cost;
}

-(void) filterBySubtype:(NSString *)subtype
{
    self.subtype = subtype;
    self.subtypes = nil;
}

-(void) filterBySubtypes:(NSSet *)subtypes
{
    self.subtype = @"";
    self.subtypes = subtypes;
}

-(void) filterByStrength:(int)strength
{
    self.strength = strength;
}

-(void) filterByAgendaPoints:(int)ap
{
    self.agendaPoints = ap;
}

-(void) filterByUniqueness:(BOOL)unique
{
    self.unique = unique;
}

-(void) filterByLimited:(BOOL)limited
{
    self.limited = limited;
}

-(void) filterByAltArt:(BOOL)altart
{
    self.altart = altart;
}

-(void) sortBy:(NRCardListSortType)sortType
{
    self.sortType = sortType;
}

-(NSMutableArray*) applyFilters
{
    NSMutableArray* filteredCards = [self.initialCards mutableCopy];
    NSMutableArray* predicates = [NSMutableArray array];
    
    if (self.faction.length > 0 && ![self.faction isEqualToString:kANY])
    {
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"factionStr LIKE[cd] %@", self.faction];
        [predicates addObject:predicate];
    }
    if (self.factions.count > 0)
    {
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"factionStr IN %@", self.factions];
        [predicates addObject:predicate];
    }
    if (self.type.length > 0 && ![self.type isEqualToString:kANY])
    {
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"typeStr LIKE[cd] %@", self.type];
        [predicates addObject:predicate];
    }
    if (self.types.count > 0)
    {
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"typeStr IN %@", self.types];
        [predicates addObject:predicate];
    }
    if (self.mu != -1)
    {
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"mu == %d", self.mu];
        [predicates addObject:predicate];
    }
    if (self.trash != -1)
    {
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"trash == %d", self.trash];
        [predicates addObject:predicate];
    }
    if (self.strength != -1)
    {
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"strength == %d", self.strength];
        [predicates addObject:predicate];
    }
    if (self.influence != -1)
    {
        if (self.faction4inf == NRFactionNone)
        {
            NSPredicate* predicate = [NSPredicate predicateWithFormat:@"influence == %d", self.influence];
            [predicates addObject:predicate];
        }
        else
        {
            NSPredicate* predicate = [NSPredicate predicateWithFormat:@"influence == %d && faction != %d", self.influence, self.faction4inf];
            [predicates addObject:predicate];
        }
    }
    if (self.set.length > 0 && ![self.set isEqualToString:kANY])
    {
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"setName LIKE[cd] %@", self.set];
        [predicates addObject:predicate];
    }
    if (self.sets.count > 0)
    {
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"setName IN %@", self.sets];
        [predicates addObject:predicate];
    }
    if (self.cost != -1)
    {
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"cost == %d || advancementCost == %d", self.cost, self.cost];
        [predicates addObject:predicate];
    }
    if (self.subtype.length > 0 && ![self.subtype isEqualToString:kANY])
    {
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"%@ IN subtypes", self.subtype];
        [predicates addObject:predicate];
    }
    if (self.subtypes.count > 0)
    {
        NSMutableArray *subPredicates = [NSMutableArray array];
        for (NSString* subtype in self.subtypes)
        {
            [subPredicates addObject:[NSPredicate predicateWithFormat:@"%@ IN subtypes", subtype]];
        }
        NSPredicate *subtypePredicate = [NSCompoundPredicate orPredicateWithSubpredicates:subPredicates];
        [predicates addObject:subtypePredicate];
    }
    if (self.agendaPoints != -1)
    {
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"agendaPoints == %d", self.agendaPoints];
        [predicates addObject:predicate];
    }
    if (self.text.length > 0)
    {
        NSPredicate* predicate;
        switch (self.searchScope)
        {
            case NRSearchAll:
                predicate = [NSPredicate predicateWithFormat:@"(name CONTAINS[cd] %@) OR (text CONTAINS[cd] %@)", self.text, self.text];
                break;
            case NRSearchName:
            {
                predicate = [NSPredicate predicateWithFormat:@"name CONTAINS[cd] %@", self.text];
                unichar ch = [self.text characterAtIndex:0];
                if (isdigit(ch))
                {
                    NSPredicate* codePredicate = [NSPredicate predicateWithFormat:@"code BEGINSWITH %@", self.text ];
                    predicate = [NSCompoundPredicate orPredicateWithSubpredicates:@[ predicate, codePredicate ]];
                }
                break;
            }
            case NRSearchText:
                predicate = [NSPredicate predicateWithFormat:@"text CONTAINS[cd] %@", self.text];
                break;
        }
        [predicates addObject:predicate];
    }
    if (self.unique)
    {
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"unique == 1"];
        [predicates addObject:predicate];
    }
    if (self.limited)
    {
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"maxPerDeck == 1"];
        [predicates addObject:predicate];
    }
    if (self.altart)
    {
        NSPredicate* predicate = [NSPredicate predicateWithFormat:@"altCard != NIL"];
        [predicates addObject:predicate];
    }

    if (predicates.count > 0)
    {
        NSPredicate* allPredicates = [NSCompoundPredicate andPredicateWithSubpredicates:predicates];
        [filteredCards filterUsingPredicate:allPredicates];
    }
    
    return filteredCards;
}

-(void) sort:(NSMutableArray*)cards
{
    [cards sortUsingComparator:^NSComparisonResult(Card* c1, Card* c2) {
        if (c1.type > c2.type)
        {
            return NSOrderedDescending;
        }
        else if (c1.type < c2.type)
        {
            return NSOrderedAscending;
        }
        
        NSComparisonResult cmp = NSOrderedSame;
        if (self.sortType == NRCardListSortFactionA_Z)
        {
            cmp = [c1.factionStr compare:c2.factionStr];
        }
        if (cmp == NSOrderedSame)
        {
            return [c1.name localizedCaseInsensitiveCompare:c2.name];
        }
        return cmp;
    }];
}

-(NSUInteger) count
{
    NSArray* arr = [self applyFilters];
    return arr.count;
}

-(TableData*) dataForTableView
{
    NSMutableArray* sections = [NSMutableArray array];
    NSMutableArray* cards = [NSMutableArray array];
    
    NSMutableArray* filteredCards = [self applyFilters];
    [self sort:filteredCards];
    NRCardType prevType = NRCardTypeNone;
    NSMutableArray* arr;
    for (Card* card in filteredCards)
    {
        if (card.type != prevType)
        {
            [sections addObject:card.typeStr];
            if (arr != nil)
            {
                [cards addObject:arr];
            }
            arr = [NSMutableArray array];
        }
        [arr addObject:card];
        prevType = card.type;
    }
    
    if (arr.count > 0)
    {
        [cards addObject:arr];
    }
    
    NSAssert(sections.count == cards.count, @"count mismatch");
    
    return [[TableData alloc] initWithSections:sections andValues:cards];
}

@end

File: NetDeck/CardList.swift
//
//  CardList.swift
//  NetDeck
//
//  Created by Gereon Steffens on 23.11.15.
//  Copyright © 2018 Gereon Steffens. All rights reserved.
//

import Foundation
import SwiftyUserDefaults

class CardList {
    private var role: Role
    private var initialCards = [Card]()
    
    private var cost = -1
    private var strength = -1
    private var mu = -1
    private var influence = -1
    private var trash = -1
    private var agendaPoints = -1
    
    private var types: Set<String>?
    private var subtype: String?
    private var subtypes: Set<String>?
    private var factions: Set<String>?
    private var sets: Set<String>?
    
    private var text: String?
    private var searchScope = CardSearchScope.all
    private var unique = false
    private var limited = false
    private var mwl = false
    
    private var faction4inf = Faction.none   // faction for influence filter
    
    private var sortType = BrowserSort.byType
    private var packUsage: PackUsage

    init(role: Role, packUsage: PackUsage, browser: Bool) {
        self.role = role
        self.packUsage = packUsage

        if browser {
            let roles = role == .none ? [ .runner, .corp ] : [ role ]
            for role in roles {
                self.initialCards.append(contentsOf: CardManager.allFor(role))
                self.initialCards.append(contentsOf: CardManager.identitiesFor(role))
            }
        } else {
            self.initialCards = CardManager.allFor(self.role)
        }

        switch self.packUsage {
        case .selected:
            self.filterDeselectedSets()
            self.addPrebuilts(includeIdentities: browser)
        case .all:
            self.filterDraft()
            self.filterRotation()
        }

        self.sortCards()
        self.clearFilters()
    }
    
    func clearFilters() {
        self.cost = -1
        self.mu = -1
        self.strength = -1
        self.trash = -1
        self.influence = -1
        self.agendaPoints = -1

        self.types = nil
        self.subtypes = nil
        self.factions = nil
        self.sets = nil
        
        self.text = ""
        self.searchScope = .all
        
        self.unique = false
        self.limited = false
        self.mwl = false
        
        self.faction4inf = .none
    }
    
    private func filterDeselectedSets() {
        let disabledPackCodes = PackManager.disabledPackCodes()
        let packPredicate = NSPredicate(format: "!(packCode in %@)", disabledPackCodes)

        self.applyPredicate(packPredicate)
    }

    private func addPrebuilts(includeIdentities: Bool) {
        for pb in Prebuilt.ownedPrebuilts {
            for var code in pb.cards.keys {
                if Defaults[.useCore2] {
                    code = Card.originalToRevised[code] ?? code
                }
                if let card = CardManager.cardBy(code) {
                    if Defaults[.rotationActive] {
                        if PackManager.rotatedPackCodes().contains(card.packCode) {
                            continue
                        }
                    }
                    if !includeIdentities && card.type == .identity {
                        continue
                    }

                    if !self.initialCards.contains(card) {
                        self.initialCards.append(card)
                    }
                }
            }
        }
    }

    private func filterRotation() {
        let rotatedPackCodes = PackManager.rotatedPackCodes()
        let predicate = NSPredicate(format: "!(packCode in %@)", rotatedPackCodes)
        self.applyPredicate(predicate)
    }
    
    func filterDraft() {
        if !Defaults[.useDraft] {
            let predicate = NSPredicate(format: "packCode != %@", PackManager.draft)
            self.applyPredicate(predicate)
        }
    }
    
    private func applyBans(_ mwl: MWL) {
        self.initialCards = self.initialCards.filter { !$0.banned(mwl) }
    }
    
    func preFilterForCorp(_ identity: Card, _ mwl: MWL) {
        if identity.faction != .neutral {
            let factions: NSArray = [ Faction.neutral.rawValue, identity.faction.rawValue ]
            let predicate = NSPredicate(format:"type != %d OR (type = %d AND faction in %@)", CardType.agenda.rawValue, CardType.agenda.rawValue, factions)
            
            self.applyPredicate(predicate)
        }
        
        if identity.code == Card.customBiotics {
            let predicate = NSPredicate(format:"faction != %d", Faction.jinteki.rawValue)
            self.applyPredicate(predicate)
        }
        
        self.applyBans(mwl)
    }
    
    func preFilterForRunner(_ identity: Card, _ mwl: MWL) {
        self.applyBans(mwl)
    }
    
    func filterByType(_ types: FilterValue) {
        self.types = types.isAny ? nil : types.strings!
    }
    
    func filterByFaction(_ factions: FilterValue) {
        self.factions = factions.isAny ? nil : factions.strings!
    }
    
    func filterBySet(_ sets: FilterValue) {
        self.sets = sets.isAny ? nil : sets.strings!
    }
    
    func filterBySubtype(_ subtypes: FilterValue) {
        self.subtypes = subtypes.isAny ? nil : subtypes.strings!
    }

    func filterByText(_ text: String) {
        self.text = text
        self.searchScope = .text
    }
    
    func filterByTextOrName(_ text: String) {
        self.text = text
        self.searchScope = .all
    }
    
    func filterByName(_ name: String) {
        self.text = name
        self.searchScope = .name
    }

    func filterByInfluence(_ influence: Int) {
        self.influence = influence
        self.faction4inf = .none
    }
    
    func filterByInfluence(_ influence: Int, forFaction faction : Faction) {
        self.influence = influence
        self.faction4inf = faction
    }
    
    func filterByMU(_ mu: Int) {
        self.mu = mu
    }
    
    func filterByTrash(_ trash: Int) {
        self.trash = trash
    }
    
    func filterByCost(_ cost: Int) {
        self.cost = cost
    }
    
    func filterByStrength(_ strength: Int) {
        self.strength = strength
    }
    
    func filterByAgendaPoints(_ ap: Int) {
        self.agendaPoints = ap
    }
    
    func filterByUniqueness(_ unique: Bool) {
        self.unique = unique
    }
    
    func filterByLimited(_ limited: Bool) {
        self.limited = limited
    }
    
    func filterByMWL(_ mwl: Bool) {
        self.mwl = mwl
    }

    func sortBy(_ sortType: BrowserSort) {
        self.sortType = sortType
        self.sortCards()
    }
    
    func applyPredicate(_ predicate: NSPredicate) {
        self.initialCards = self.initialCards.filter { predicate.evaluate(with: $0) }
    }
    
    private func applyFilters() -> [Card] {
        var filteredCards = self.initialCards
        var predicates = [NSPredicate]()

        if let f = self.factions, f.count > 0 {
            let predicate = NSPredicate(format:"factionStr IN %@", f)
            predicates.append(predicate)
        }
        if let t = self.types, t.count > 0 {
            let predicate = NSPredicate(format:"typeStr IN %@", t)
            predicates.append(predicate)
        }
        if let s = self.sets, s.count > 0 {
            let predicate = NSPredicate(format:"packName IN %@", s)
            predicates.append(predicate)
        }
        if let s = self.subtypes, s.count > 0 {
            var subPredicates = [NSPredicate]()
            for subtype in s {
                subPredicates.append(NSPredicate(format:"%@ IN subtypes", subtype))
            }
            let subtypePredicate = NSCompoundPredicate(orPredicateWithSubpredicates:subPredicates)
            predicates.append(subtypePredicate)
        }

        if self.mu != -1 {
            let predicate = NSPredicate(format:"mu == %d", self.mu)
            predicates.append(predicate)
        }
        if self.trash != -1 {
            let predicate = NSPredicate(format:"trash == %d", self.trash)
            predicates.append(predicate)
        }
        if self.strength != -1 {
            let predicate = NSPredicate(format:"strength == %d", self.strength)
            predicates.append(predicate)
        }
        if self.influence != -1 {
            if self.faction4inf == .none {
                let predicate = NSPredicate(format:"influence == %d", self.influence)
                predicates.append(predicate)
            } else {
                let predicate = NSPredicate(format:"influence == %d && faction != %d", self.influence, self.faction4inf.rawValue)
                predicates.append(predicate)
            }
        }
        if self.cost != -1 {
            let predicate = NSPredicate(format:"cost == %d || advancementCost == %d", self.cost, self.cost)
            predicates.append(predicate)
        }
        if self.agendaPoints != -1 {
            let predicate = NSPredicate(format:"agendaPoints == %d", self.agendaPoints)
            predicates.append(predicate)
        }
        
        if let text = self.text, text.count > 0 {
            var predicate: NSPredicate
            switch (self.searchScope) {
            case .all:
                predicate = NSPredicate(format:"(name CONTAINS[cd] %@) OR (englishName CONTAINS[cd] %@) OR (text CONTAINS[cd] %@) or (ANY aliases CONTAINS[cd] %@)",
                                        text, text, text, text)
            case .name:
                predicate = NSPredicate(format:"(name CONTAINS[cd] %@) OR (englishName CONTAINS[cd] %@) OR (ANY aliases CONTAINS[cd] %@)",
                                        text, text, text)
                let ch = text[text.startIndex]
                if ch >= "0" && ch <= "9" {
                    let codePredicate = NSPredicate(format:"code BEGINSWITH %@", text)
                    predicate = NSCompoundPredicate(orPredicateWithSubpredicates:[ predicate, codePredicate ])
                }
            case .text:
                predicate = NSPredicate(format:"text CONTAINS[cd] %@", text)
            }
            predicates.append(predicate)
        }
        
        if self.unique {
            let predicate = NSPredicate(format:"unique == 1")
            predicates.append(predicate)
        }
        if self.limited {
            let predicate = NSPredicate(format:"type != %d AND maxPerDeck == 1", CardType.identity.rawValue)
            predicates.append(predicate)
        }
        
        if predicates.count > 0 {
            let allPredicates = NSCompoundPredicate(andPredicateWithSubpredicates:predicates)
            filteredCards = filteredCards.filter { allPredicates.evaluate(with: $0) }
        }
        
        if self.mwl {
            let mwl = Defaults[.defaultMWL]
            if mwl != .none {
                filteredCards = filteredCards.filter { $0.mwlPenalty(mwl) > 0 || $0.banned(mwl) || $0.restricted(mwl) }
            }
        }

        return filteredCards
    }
    
    private func sortCards() {
        print("sorting card list")
        let start = Date.timeIntervalSinceReferenceDate
        self.initialCards.sort { c1, c2 in
            switch self.sortType {
            case .byType, .byTypeFaction:
                if c1.type.rawValue < c2.type.rawValue { return true }
                if c1.type.rawValue > c2.type.rawValue { return false }
            case .byFaction:
                if c1.factionStr < c2.factionStr { return true }
                if c1.factionStr > c2.factionStr { return false }
            case .bySet, .bySetType, .bySetFaction:
                if c1.packNumber < c2.packNumber { return true }
                if c1.packNumber > c2.packNumber { return false }
            case .bySetNumber:
                if c1.packNumber < c2.packNumber { return true }
                if c1.packNumber > c2.packNumber { return false }
            case .byCost:
                if c1.cost < c2.cost { return true }
                if c1.cost > c2.cost { return false }
            case .byStrength:
                if c1.strength < c2.strength { return true }
                if c1.strength > c2.strength { return false }
            }
            
            switch self.sortType {
            case .byTypeFaction, .bySetFaction:
                if c1.factionStr < c2.factionStr { return true }
                if c1.factionStr > c2.factionStr { return false }
            case .bySetType:
                if c1.type.rawValue < c2.type.rawValue { return true }
                if c1.type.rawValue > c2.type.rawValue { return false }
            case .bySetNumber:
                if c1.number < c2.number { return true }
                if c1.number > c2.number { return false }
            default: break
            }
            
            return c1.foldedName < c2.foldedName
        }
        let elapsed = Date.timeIntervalSinceReferenceDate - start
        print("took \(elapsed)s")
    }
    
    func count() -> Int {
        let arr = self.applyFilters()
        return arr.count
    }
    
    func allCards() -> [Card] {
        return self.applyFilters()
    }
    
    func dataForTableView() -> TableData<Card> {
        let filteredCards = self.applyFilters()
        
        var sections = [String]()
        var cards = [[Card]]()
        var current = [Card]()
        for card in filteredCards {
            if current.isEmpty || self.section(card) == self.section(current[0]) {
                current.append(card)
            } else {
                sections.append(self.section(current[0]))
                cards.append(current)
                current = [card]
            }
        }
        if !current.isEmpty {
            sections.append(self.section(current[0]))
            cards.append(current)
        }

        return TableData(sections: sections, values: cards)
    }
    
    private func section(_ card: Card) -> String {
        switch self.sortType {
        case .byType, .byTypeFaction:
            return card.typeStr
        case .byFaction:
            return card.factionStr
        case .bySet, .bySetType, .bySetNumber, .bySetFaction:
            return card.packName
        case .byCost:
            return "Cost".localized() + " \(card.costString)"
        case .byStrength:
            return "Strength".localized() + " \(card.strengthString)"
        }
    }
    
}
